File: listaenlazada/Lista_Enlazada.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

let head = null;
const prompt = require("prompt-sync")();

function beginInsert() {
    let item = parseInt(prompt("Ingrese valor: "));
    let node = new Node(item);
    node.next = head;
    head = node;
    console.log("Nodo insertado");
}

function lastInsert() {
    let item = parseInt(prompt("Ingrese valor: "));
    let node = new Node(item);

    if (!head) {
        head = node;
    } else {
        let temp = head;
        while (temp.next)
            temp = temp.next;
        temp.next = node;
    }
    console.log("Nodo insertado");
}

function randomInsert() {
    let item = parseInt(prompt("Ingrese valor: "));
    let loc = parseInt(prompt("Ingrese posición: "));

    let temp = head;
    for (let i = 0; i < loc; i++) {
        if (!temp) {
            console.log("No se puede insertar");
            return;
        }
        temp = temp.next;
    }

    let node = new Node(item);
    node.next = temp.next;
    temp.next = node;

    console.log("Nodo insertado");
}

function beginDelete() {
    if (!head) return console.log("Lista vacía");
    head = head.next;
    console.log("Nodo eliminado");
}

function lastDelete() {
    if (!head) return console.log("Lista vacía");
    if (!head.next) {
        head = null;
        return console.log("Nodo eliminado");
    }

    let temp = head, prev = null;
    while (temp.next) {
        prev = temp;
        temp = temp.next;
    }
    prev.next = null;
    console.log("Nodo eliminado");
}

function randomDelete() {
    let loc = parseInt(prompt("Ingrese posición: "));

    if (!head) return console.log("Lista vacía");

    if (loc === 0) {
        head = head.next;
        return console.log("Nodo eliminado");
    }

    let temp = head, prev = null;
    for (let i = 0; i < loc; i++) {
        prev = temp;
        temp = temp.next;
        if (!temp) return console.log("No se puede eliminar");
    }

    prev.next = temp.next;
    console.log("Nodo eliminado");
}

function display() {
    if (!head) return console.log("Lista vacía");

    let temp = head;
    let out = "";
    while (temp) {
        out += temp.data + "->";
        temp = temp.next;
    }
    console.log(out + "NULL");
}

function search() {
    let item = parseInt(prompt("Ingrese valor: "));
    let temp = head, pos = 0;

    while (temp) {
        if (temp.data === item) {
            console.log("Encontrado en posición " + pos);
            return;
        }
        temp = temp.next;
        pos++;
    }
    console.log("No encontrado");
}

function menu() {
    let choice = 0;
    while (choice !== 9) {
        console.log(`
*******MENU*******
1. Insert beginning
2. Insert last
3. Insert random
4. Delete beginning
5. Delete last
6. Delete random
7. Display
8. Search
9. Exit
`);
        choice = parseInt(prompt("Choice: "));

        switch (choice) {
            case 1: beginInsert(); break;
            case 2: lastInsert(); break;
            case 3: randomInsert(); break;
            case 4: beginDelete(); break;
            case 5: lastDelete(); break;
            case 6: randomDelete(); break;
            case 7: display(); break;
            case 8: search(); break;
        }
    }
}

menu();

File: listaenlazada/Lista_Enlazada.cs
using System;

class Node {
    public int Data;
    public Node Next;
}

class Program {
    static Node head;

    static void Main() {
        int choice = 0;
        while (choice != 9) {
            Console.WriteLine("\n*********Main Menu*********");
            Console.WriteLine("1. Insert at beginning");
            Console.WriteLine("2. Insert at last");
            Console.WriteLine("3. Insert at random location");
            Console.WriteLine("4. Delete from beginning");
            Console.WriteLine("5. Delete from last");
            Console.WriteLine("6. Delete from random location");
            Console.WriteLine("7. Display");
            Console.WriteLine("8. Search");
            Console.WriteLine("9. Exit");
            Console.Write("Enter your choice: ");

            choice = int.Parse(Console.ReadLine());

            switch (choice) {
                case 1: BeginInsert(); break;
                case 2: LastInsert(); break;
                case 3: RandomInsert(); break;
                case 4: BeginDelete(); break;
                case 5: LastDelete(); break;
                case 6: RandomDelete(); break;
                case 7: Display(); break;
                case 8: Search(); break;
            }
        }
    }

    static void BeginInsert() {
        Console.Write("Ingrese valor: ");
        int item = int.Parse(Console.ReadLine());
        Node newNode = new Node { Data = item, Next = head };
        head = newNode;
        Console.WriteLine("Nodo insertado");
    }

    static void LastInsert() {
        Console.Write("Ingrese valor: ");
        int item = int.Parse(Console.ReadLine());
        Node newNode = new Node { Data = item };

        if (head == null) {
            head = newNode;
        } else {
            Node temp = head;
            while (temp.Next != null)
                temp = temp.Next;
            temp.Next = newNode;
        }
        Console.WriteLine("Nodo insertado");
    }

    static void RandomInsert() {
        Console.Write("Ingrese valor: ");
        int item = int.Parse(Console.ReadLine());
        Console.Write("Ingrese posicion: ");
        int loc = int.Parse(Console.ReadLine());

        Node temp = head;
        for (int i = 0; i < loc; i++) {
            if (temp == null) {
                Console.WriteLine("No se puede insertar");
                return;
            }
            temp = temp.Next;
        }

        Node newNode = new Node { Data = item, Next = temp.Next };
        temp.Next = newNode;
        Console.WriteLine("Nodo insertado");
    }

    static void BeginDelete() {
        if (head == null) {
            Console.WriteLine("Lista vacía");
            return;
        }
        head = head.Next;
        Console.WriteLine("Nodo eliminado");
    }

    static void LastDelete() {
        if (head == null) { Console.WriteLine("Lista vacía"); return; }
        if (head.Next == null) { head = null; Console.WriteLine("Nodo eliminado"); return; }

        Node temp = head, prev = null;
        while (temp.Next != null) {
            prev = temp;
            temp = temp.Next;
        }
        prev.Next = null;
        Console.WriteLine("Nodo eliminado");
    }

    static void RandomDelete() {
        Console.Write("Ingrese posición: ");
        int loc = int.Parse(Console.ReadLine());

        if (head == null) { Console.WriteLine("Lista vacía"); return; }

        if (loc == 0) {
            head = head.Next;
            Console.WriteLine("Nodo eliminado");
            return;
        }

        Node temp = head, prev = null;
        for (int i = 0; i < loc; i++) {
            prev = temp;
            temp = temp.Next;
            if (temp == null) {
                Console.WriteLine("No se puede eliminar");
                return;
            }
        }

        prev.Next = temp.Next;
        Console.WriteLine("Nodo eliminado");
    }

    static void Display() {
        if (head == null) {
            Console.WriteLine("Lista vacía");
            return;
        }

        Node temp = head;
        while (temp != null) {
            Console.Write(temp.Data + "->");
            temp = temp.Next;
        }
        Console.WriteLine("NULL");
    }

    static void Search() {
        Console.Write("Ingrese valor a buscar: ");
        int item = int.Parse(Console.ReadLine());

        Node temp = head;
        int pos = 0;
        while (temp != null) {
            if (temp.Data == item) {
                Console.WriteLine($"Elemento {item} encontrado en {pos}");
                return;
            }
            temp = temp.Next;
            pos++;
        }
        Console.WriteLine("Elemento no encontrado");
    }
}
